File: src.js/data/savefile-expanded/sections/Storage.js
.pragma library

.import "../fragments/PokemonBox.js" as PokemonBox_
const PokemonBox = PokemonBox_.PokemonBox;

class Storage {
  constructor(saveFile) {
    this.curBox = 0;
    this.changedBoxesBefore = false;
    this.boxItems = [];
    this.pokemonBoxes = [[]];

    if (saveFile !== undefined)
      this.load(saveFile);
  }

  load(saveFile) {
    let curBox = this.curBox = (saveFile.getByte(0x284C) & 0b01111111);

    // All the Pokemon data is uninitialized and unformatted / random
    // garbage if this is false
    // So if this is false then don't load any box data
    this.changedBoxesBefore = saveFile.getBit(0x284C, 0x1, 7);

    const it = saveFile.iterator.offsetTo(0x27E7);
    this.boxItems = [];
    for (let i = 0; i < saveFile.getByte(0x27E6) && i < 50; i++) {
      this.boxItems.push({
                           id: it.getByte(),
                           amount: it.getByte()
                         });
    }

    // Create Parent of all boxes
    this.pokemonBoxes = [];

    // Load boxes 1-6
    this.loadBoxes(saveFile, 0x4000, 0);

    // Load boxes 7-12
    this.loadBoxes(saveFile, 0x6000, 1);

    // Overwrite box data of the current box with current box data
    // Pokemon Red and Blue save a copy of the current box locally in Bank 1
    // All box changes go to this box, in other words
    // If the current box is box number 1, then the actual box number 1
    // reflects all changes up until the box switch and none after. Box 1,
    // in this example, will only be updated to reflect all newest changes
    // when switching boxes

    // Wipe current box data and re-insert from scratch latest box data for
    // that box

    // Also force load the current box
    this.pokemonBoxes[curBox] = [];
    this.loadBox(saveFile, curBox, 0x30C0, true);
  }

  // Loads 6 boxes in a bank offset adding to the existing boxes
  loadBoxes(saveFile, boxesOffset, boxSet) {

    // i = Box Number within set/bank (0-5)
    // j = Box Number within all boxes (0-11)
    for (let i = 0, j = boxSet * 6; j < (boxSet * 6) + 6; i++ , j++) {
      // Skip current box (load only from cached box)
      if (j == this.curBox)
        continue;

      this.pokemonBoxes[j] = [];
      this.loadBox(
        saveFile,
        j,
        (i * 0x462) + boxesOffset
        )
    }
  }

  // Loads a pokemon box # from given offset, won't load anything unless
  // data is pre-formatted or forced
  loadBox(saveFile, boxInd, boxOffset, forceLoad = false) {
    if (!this.changedBoxesBefore && !forceLoad)
      return;

    for (let i = 0; i < saveFile.getByte(boxOffset) && i < 20; i++) {
      this.pokemonBoxes[boxInd].push(new PokemonBox(
                                       saveFile,
                                       boxOffset + 0x16,
                                       boxOffset + 0x386,
                                       boxOffset + 0x2AA,
                                       i));
    }
  }

  save(saveFile) {
    // Save current box details
    saveFile.setByte(0x284C, this.curBox);
    saveFile.setBit(0x284C, 1, 7, this.changedBoxesBefore);

    // Save all box items
    const it = saveFile.iterator.offsetTo(0x27E6);
    it.setByte(this.boxItems.length);
    for (let i = 0; i < this.boxItems.length && i < 50; i++) {
      it.setByte(this.boxItems[i].id);
      it.setByte(this.boxItems[i].amount);
    }
    it.setByte(0xFF);

    // Save boxes 1-6
    this.saveBoxes(
      saveFile,
      0x4000,
      0
      );

    // Save boxes 7-12
    this.saveBoxes(
      saveFile,
      0x6000,
      1
      );

    // Save current box data to current box cache
    this.saveBox(
      saveFile,
      this.curBox,
      0x30C0,
      true
      );
  }

  saveBox(saveFile, boxInd, boxOffset, force = false) {
    // Don't write anything unless changedBoxesBefore is set unless forced
    if (!this.changedBoxesBefore && !force)
      return;

    saveFile.setByte(boxOffset, this.pokemonBoxes[boxInd].length);
    for (let i = 0; i < this.pokemonBoxes[boxInd].length && i < 20; i++) {
      this.pokemonBoxes[boxInd][i].save(
        saveFile,
        boxOffset + 0x16,
        boxOffset + 0x1,
        boxOffset + 0x386,
        boxOffset + 0x2AA,
        i
        );
    }

    // Mark end of species list if not full box
    if(this.pokemonBoxes[boxInd].length >= 20)
      return;

    let speciesOffset = boxOffset + 1 + this.pokemonBoxes[boxInd].length;
    saveFile.setByte(speciesOffset, 0xFF);
  }

  saveBoxes(saveFile, boxesOffset, boxSet) {
    // i = Box Number within set/bank (0-5)
    // j = Box Number within all boxes (0-11)
    for (let i = 0, j = boxSet * 6; j < (boxSet * 6) + 6; i++ , j++) {
      // Skip current box (save only to cached box)
      if (j == this.curBox)
        continue;

      this.saveBox(
        saveFile,
        j,
        (i * 0x462) + boxesOffset
        )
    }
  }
}

File: src/data/file/expanded/storage.cpp
/*
  * Copyright 2020 June Hanabi
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *   http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

#include "./storage.h"
#include "./fragments/pokemonstorageset.h"
#include "./fragments/pokemonstoragebox.h"
#include "./fragments/itemstoragebox.h"
#include "../savefile.h"
#include "../savefiletoolset.h"
#include "../savefileiterator.h"

Storage::Storage(SaveFile* saveFile)
{
  items = new ItemStorageBox(false, 50); // Max 50 items, Mark this is storage

  for(var8 i = 0; i < maxPokemonStorageSets; i++)
    pokemon[i] = new PokemonStorageSet;

  load(saveFile);
}

Storage::~Storage()
{
  delete items;

  for(var8 i = 0; i < maxPokemonStorageSets; i++)
    delete pokemon[i];
}

int Storage::setCount()
{
  return maxPokemonStorageSets;
}

PokemonStorageSet* Storage::setAt(int ind)
{
  return pokemon[ind];
}

void Storage::setSwap(int from, int to)
{
  auto eFrom = pokemon[from];
  auto eTo = pokemon[to];

  pokemon[to] = eFrom;
  pokemon[from] = eTo;

  pokemonChanged();
}

int Storage::boxCount()
{
  return maxPokemonBoxes;
}

PokemonStorageBox* Storage::boxAt(int ind)
{
  bool curSetB = ind >= setMaxBoxes;
  var8 cur = ind; // Get the current box
  if(curSetB)
    cur -= setMaxBoxes; // Offset if it's in set b

  return pokemon[curSetB ? 1 : 0]->boxes[cur];
}

void Storage::boxSwap(int from, int to)
{
  // Get correct set, index within set, and pointer for from box
  bool fromSetB = from >= setMaxBoxes;
  var8 fromBox = from; // Get the current box
  if(fromSetB)
    fromBox -= setMaxBoxes;

  auto fromBoxData = pokemon[fromSetB ? 1 : 0]->boxes[from];

  // Get correct set, index within set, and pointer for to box
  bool toSetB = to >= setMaxBoxes;
  var8 toBox = to; // Get the current box
  if(toSetB)
    toBox -= setMaxBoxes;

  auto toBoxData = pokemon[toSetB ? 1 : 0]->boxes[to];

  // Make the swap ignoring set boundraries
  pokemon[fromSetB ? 1 : 0]->boxes[from] = toBoxData;
  pokemon[toSetB ? 1 : 0]->boxes[to] = fromBoxData;

  // Contents within the sets have changed by the set
  pokemonChanged();
}

PokemonStorageBox* Storage::freeSpace()
{
  if(!boxesFormatted) {
    if(boxAt(curBox)->pokemonCount() < boxAt(curBox)->pokemonMax())
      return boxAt(curBox);
    else
      return nullptr;
  }

  for(int i = 0; i < boxCount(); i++) {
    if(boxAt(i)->pokemonCount() < boxAt(i)->pokemonMax())
      return boxAt(curBox);;
  }

  return nullptr;
}

bool Storage::depositPokemon(PokemonBox* pokemon)
{
  auto storage = freeSpace();
  if(storage == nullptr)
    return false;

  storage->pokemon.append(pokemon);
  storage->pokemonChanged();
  return true;
}

void Storage::load(SaveFile* saveFile)
{
  reset();

  if(saveFile == nullptr)
    return;

  auto toolset = saveFile->toolset;

  // Now Pokemon operates off a cached version of a box, so when switching boxes
  // it saves the box from the cache to storage and then loads the new box into
  // cache. All changes work off of the cache and only get saved back to storage
  // when switching boxes. It's imperative we replicate this.

  // Current Box, last bit represents whether the other boxes are formatted or
  // not. If they aren't formatted then don't load them as they contain garbage
  // information
  curBox = (toolset->getByte(0x284C) & 0b01111111);
  curBoxChanged();

  boxesFormatted = toolset->getBit(0x284C, 0x1, 7);
  boxesFormattedChanged();

  // Load Items and Boxes 1-12 in 2 sets of 1-6
  items->load(saveFile, 0x27E6);

  // Determine how to play out current box loading
  // We only want to load all but current box
  bool curSetB = curBox >= setMaxBoxes; // Which set is the current box in
  var8 cur = curBox; // Get the current box
  if(curSetB)
    cur -= setMaxBoxes; // Offset if it's in set b

  // Load all of these and skip current box in correct set only if the boxes
  // have been formatted
  if(boxesFormatted) {
    pokemon[0]->load(saveFile, 0x4000, curSetB ? -1 : cur);
    pokemon[1]->load(saveFile, 0x6000, curSetB ? cur : -1);
  }

  // Load cached box to current box data regardless of boxes being formatted
  // or not
  pokemon[curSetB ? 1 : 0]->loadSpecific(saveFile, 0x30C0, cur);
}

void Storage::save(SaveFile* saveFile)
{
  auto toolset = saveFile->toolset;

  // Save current box details
  toolset->setByte(0x284C, curBox);
  toolset->setBit(0x284C, 1, 7, boxesFormatted);

  items->save(saveFile, 0x27E6);

  bool curSetB = curBox >= setMaxBoxes; // Which set is the current box in
  var8 cur = curBox; // Get the current box
  if(curSetB)
    cur -= setMaxBoxes; // Offset if it's in set b

  // Load all of these and skip current box in correct set only if the boxes
  // have been formatted
  if(boxesFormatted) {
    pokemon[0]->save(saveFile, 0x4000, curSetB ? -1 : cur);
    pokemon[1]->save(saveFile, 0x6000, curSetB ? cur : -1);
  }

  // Load cached box to current box data regardless of boxes being formatted
  // or not
  pokemon[curSetB ? 1 : 0]->saveSpecific(saveFile, 0x30C0, cur);
}

void Storage::reset()
{
  curBox = 0;
  curBoxChanged();

  boxesFormatted = false;
  boxesFormattedChanged();

  items->reset();

  for(var8 i = 0; i < maxPokemonStorageSets; i++)
    pokemon[i]->reset();
}

void Storage::randomize(PlayerBasics* basics)
{
  reset();
  items->randomize();

  for(var8 i = 0; i < maxPokemonStorageSets; i++)
    pokemon[i]->randomize(basics);
}
